### PR TITLE
ampctl and other stuff

### PR DESCRIPTION
**Pull request by mdblack98**

### CODE REVIEW

**Comment by N0NB** [comment]
I am seeing 'make distcheck' fail:

```
make[3]: Entering directory '/home/nate/git/hlb/amplifiers/elecraft'
make  distdir-am
make[4]: Entering directory '/home/nate/git/hlb/amplifiers/elecraft'
make[4]: *** No rule to make target 'README.elecraft', needed by 'distdir-am'.  Stop.
make[4]: Leaving directory '/home/nate/git/hlb/amplifiers/elecraft'
make[3]: *** [Makefile:545: distdir] Error 2
make[3]: Leaving directory '/home/nate/git/hlb/amplifiers/elecraft'
make[2]: *** [Makefile:713: distdir-am] Error 1
```
It looks like the `README.elecraft` file didn't make it in.  If you have one, add another patch or I can apply a patch to the `amplifiers/elecraft/Makefile.am`.

**Comment by mdblack98** [comment]
I updated the pull request.
Mike

 

    On Tuesday, June 25, 2019, 07:11:13 PM CDT, Nate Bargmann <notifications@github.com> wrote:  
 
 
I am seeing 'make distcheck' fail:
make[3]: Entering directory '/home/nate/git/hlb/amplifiers/elecraft'
make  distdir-am
make[4]: Entering directory '/home/nate/git/hlb/amplifiers/elecraft'
make[4]: *** No rule to make target 'README.elecraft', needed by 'distdir-am'.  Stop.
make[4]: Leaving directory '/home/nate/git/hlb/amplifiers/elecraft'
make[3]: *** [Makefile:545: distdir] Error 2
make[3]: Leaving directory '/home/nate/git/hlb/amplifiers/elecraft'
make[2]: *** [Makefile:713: distdir-am] Error 1

It looks like the README.elecraft file didn't make it in. If you have one, add another patch or I can apply a patch to the amplifiers/elecraft/Makefile.am.

File: amplifiers/elecraft/make.log [hunk 1]
```diff
@@ -0,0 +1,34 @@
+  CC       kpa.lo
```
Comment (mikaelnousiainen) [inline amplifiers/elecraft/make.log]: This file should probably be deleted :)

File: include/hamlib/ampclass.h [hunk 2]
```diff
@@ -0,0 +1,65 @@
+/*
+ *  Hamlib C++ bindings - rotator API header
```
Comment (mikaelnousiainen) [inline include/hamlib/ampclass.h]: rotator?

File: include/hamlib/ampclass.h [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+ */
+
+#ifndef _ROTCLASS_H
+#define _ROTCLASS_H 1
+
+#include <hamlib/rotator.h>
+
+
+
+class BACKEND_IMPEXP Rotator
```
Comment (mikaelnousiainen) [inline include/hamlib/ampclass.h]: More rotator-related stuff in `ampclass.h`? Please review the whole file...

File: src/rig.c [hunk 4]
```diff
@@ -1962,10 +1962,10 @@ int HAMLIB_API rig_get_ptt(RIG *rig, vfo_t vfo, ptt_t *ptt)
         return par_ptt_get(&rig->state.pttport, ptt);
 
     case RIG_PTT_CM108:
-        if (caps->get_ptt)
-        {
-            return caps->get_ptt(rig, vfo, ptt);
-        }
+        //if (caps->get_ptt)
```
Comment (mikaelnousiainen) [inline src/rig.c]: There should be no reason to leave commented-out code here...

File: src/amp_conf.c [hunk 5]
```diff
@@ -0,0 +1,695 @@
...
+    "Minimum amplifier azimuth in degrees",
+    "-180", RIG_CONF_NUMERIC, { .n = { -360, 360, .001 } }
+  },
+  {
+    TOK_MAX_AZ, "max_az", "Maximum azimuth",
+    "Maximum amplifier azimuth in degrees",
+    "180", RIG_CONF_NUMERIC, { .n = { -360, 360, .001 } }
+  },
+  {
+    TOK_MIN_EL, "min_el", "Minimum elevation",
```
Comment (mikaelnousiainen) [inline src/amp_conf.c]: Rotator-related copy-paste here too...

File: src/Makefile.am [hunk 6]
```diff
@@ -4,15 +4,16 @@ RIGSRC = rig.c serial.c serial.h misc.c misc.h register.c register.h event.c \
 	event.h cal.c cal.h conf.c tones.c tones.h rotator.c locator.c rot_reg.c \
 	rot_conf.c rot_conf.h iofunc.c iofunc.h ext.c mem.c settings.c \
 	parallel.c parallel.h usb_port.c usb_port.h debug.c network.c network.h \
-	cm108.c cm108.h gpio.c gpio.h idx_builtin.h token.h par_nt.h microham.c microham.h
+	cm108.c cm108.h gpio.c gpio.h idx_builtin.h token.h par_nt.h microham.c microham.h \
+  amplifier.c amp_reg.c amp_conf.c amp_conf.h extamp.c
```
Comment (mikaelnousiainen) [inline src/Makefile.am]: Formatting issues. I wonder if we should simply do a separate PR that fixes formatting for all files before 4.0 release? We do have clearly documented rules for formatting anyway.

File: include/hamlib/amplifier.h [hunk 7]
```diff
@@ -0,0 +1,475 @@
...
+ *  \def AMP_MOVE_UP
+ *  \brief A macro that returns the flag for the \b UP direction.
+ *
+ *  This macro defines the value of the \b UP direction which can be
+ *  used with the amp_move() function.
+ *
+ *  \sa amp_move(), AMP_MOVE_DOWN, AMP_MOVE_LEFT, AMP_MOVE_CCW,
+ *      AMP_MOVE_RIGHT, AMP_MOVE_CW
+ */
+#define AMP_MOVE_UP         (1<<1)
```
Comment (mikaelnousiainen) [inline include/hamlib/amplifier.h]: `AMP_MOVE_UP/DOWN`? :) More rotator-related copy-paste to be removed.

File: include/hamlib/amplifier.h [hunk 8]
```diff
@@ -0,0 +1,475 @@
...
+  int (*set_conf)(AMP *amp, token_t token, const char *val);
+  int (*get_conf)(AMP *amp, token_t token, char *val);
+
+  /*
+   *  General API commands, from most primitive to least.. :()
+   *  List Set/Get functions pairs
+   */
+
+  int (*reset)(AMP *amp, amp_reset_t reset);
+  int (*get_level)(AMP *amp, setting_t level, value_t *val);
```
Comment (mikaelnousiainen) [inline include/hamlib/amplifier.h]: While it may be true that Elecraft amps do not support setting any additional levels, I'd like to see the API support `set_level` and especially `set_ext_level` for amps too. It's quite easy to add them now for a new API!

Comment (mikaelnousiainen) [reply include/hamlib/amplifier.h]: Also, `get_func`/`set_func` would be useful just to be complete.

File: include/hamlib/amplifier.h [hunk 9]
```diff
@@ -0,0 +1,475 @@
...
+  /*
+   * non overridable fields, internal use
+   */
+  hamlib_port_t ampport;  /*!< Amplifier port (internal use). */
+
+  int comm_state;         /*!< Comm port state, opened/closed. */
+  rig_ptr_t priv;         /*!< Pointer to private amplifier state data. */
+  rig_ptr_t obj;          /*!< Internal use by hamlib++ for event handling. */
+
+  setting_t has_get_level;
```
Comment (mikaelnousiainen) [inline include/hamlib/amplifier.h]: Possible add set-levels and set/get/conf here too?

File: dummy/amp_dummy.c [hunk 10]
```diff
@@ -0,0 +1,374 @@
...
+  struct dummy_amp_priv_data *priv = (struct dummy_amp_priv_data *)
+                                     amp->state.priv;
+
+  *status = priv->powerstat;
+  rig_debug(RIG_DEBUG_VERBOSE, "%s called\n", __FUNCTION__);
+
+  return RIG_OK;
+}
+
+#if 0
```
Comment (mikaelnousiainen) [inline dummy/amp_dummy.c]: Why is this piece of code excluded?

File: amplifiers/elecraft/kpa1500.c [hunk 11]
```diff
@@ -0,0 +1,207 @@
...
+
+  return RIG_OK;
+}
+
+
+/*
+ *
+ */
+
+#if 0
```
Comment (mikaelnousiainen) [inline amplifiers/elecraft/kpa1500.c]: More excluded code...?

File: amplifiers/elecraft/kpa.c [hunk 12]
```diff
@@ -0,0 +1,593 @@
...
+ *
+ */
+
+#include <stdlib.h>
+#include <string.h>
+#include "kpa.h"
+
+struct fault_list
+{
+  int code;
```
Comment (mikaelnousiainen) [inline amplifiers/elecraft/kpa.c]: Formatting issues. Much of the new code uses 2 spaces instead of 4. Also, the braces with if/else/while statements do not follow the documented formatting rules. We should automate the formatting process, so we could do it with a simple script for every PR before merged.
